File: customercenter/src/main/java/cafeteria/MypageViewHandler.java
package cafeteria;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cloud.stream.annotation.StreamListener;
import org.springframework.messaging.handler.annotation.Payload;
import org.springframework.stereotype.Service;

import cafeteria.config.kafka.KafkaProcessor;
import cafeteria.external.KakaoMessage;
import cafeteria.external.KakaoService;

@Service
public class MypageViewHandler {

	@Autowired
	private KakaoService kakaoService;
	
    @Autowired
    private MypageRepository mypageRepository;

    @StreamListener(KafkaProcessor.INPUT)
    public void whenOrdered_then_CREATE_1 (@Payload Ordered ordered) {
        try {
            if (ordered.isMe()) {
                // view 객체 생성
                Mypage mypage = new Mypage();
                // view 객체에 이벤트의 Value 를 set 함
                mypage.setOrderId(ordered.getId());
                mypage.setProductName(ordered.getProductName());
                mypage.setQty(ordered.getQty());
                mypage.setAmt(ordered.getAmt());
                mypage.setStatus(ordered.getStatus());
                // view 레파지 토리에 save
                mypageRepository.save(mypage);
                
                KakaoMessage message = new KakaoMessage();
                message.setPhoneNumber(ordered.getPhoneNumber());
                message.setMessage(new StringBuffer("Your Order is ").append(ordered.getStatus()).toString());
                kakaoService.sendKakao(message);
            }
        }catch (Exception e){
            e.printStackTrace();
        }
    }


    @StreamListener(KafkaProcessor.INPUT)
    public void whenOrderCanceled_then_UPDATE_1(@Payload OrderCanceled orderCanceled) {
        try {
            if (orderCanceled.isMe()) {
                // view 객체 조회
                List<Mypage> pagies = mypageRepository.findByOrderId(orderCanceled.getId());
                for(Mypage page  : pagies){
                    page.setStatus(orderCanceled.getStatus());
                	mypageRepository.save(page);
                }
                
                KakaoMessage message = new KakaoMessage();
                message.setPhoneNumber(orderCanceled.getPhoneNumber());
                message.setMessage(new StringBuffer("Your Order is ").append(orderCanceled.getStatus()).toString());
                kakaoService.sendKakao(message);
            }
        }catch (Exception e){
            e.printStackTrace();
        }
    }
    
    @StreamListener(KafkaProcessor.INPUT)
    public void whenReceipted_then_UPDATE_3(@Payload Receipted receipted) {
        try {
            if (receipted.isMe()) {
                // view 객체 조회
            	List<Mypage> pagies = mypageRepository.findByOrderId(receipted.getOrderId());
            	for(Mypage page  : pagies){
                    // view 객체에 이벤트의 eventDirectValue 를 set 함
            		page.setStatus(receipted.getStatus());
                    // view 레파지 토리에 save
                	mypageRepository.save(page);
                }
            	
            	KakaoMessage message = new KakaoMessage();
                message.setPhoneNumber(receipted.getPhoneNumber());
                message.setMessage(new StringBuffer("Your Order is ").append(receipted.getStatus()).toString());
                kakaoService.sendKakao(message);
            }
        }catch (Exception e){
            e.printStackTrace();
        }
    }
    @StreamListener(KafkaProcessor.INPUT)
    public void whenMade_then_UPDATE_4(@Payload Made made) {
        try {
            if (made.isMe()) {
                // view 객체 조회
            	List<Mypage> pagies = mypageRepository.findByOrderId(made.getOrderId());
            	for(Mypage page  : pagies){
                    // view 객체에 이벤트의 eventDirectValue 를 set 함
            		page.setStatus(made.getStatus());
                    // view 레파지 토리에 save
                	mypageRepository.save(page);
                }
            	
            	KakaoMessage message = new KakaoMessage();
                message.setPhoneNumber(made.getPhoneNumber());
                message.setMessage(new StringBuffer("Your Order is ").append(made.getStatus()).append("\nCome and Take it, Please!").toString());
                kakaoService.sendKakao(message);
            }
        }catch (Exception e){
            e.printStackTrace();
        }
    }

}

File: customercenter/src/main/scala/cafeteria/MypageViewHandler.scala
package cafeteria

import org.springframework.stereotype.Service
import org.springframework.beans.factory.annotation.Autowired
import cafeteria.external.KakaoService
import org.springframework.cloud.stream.annotation.StreamListener
import org.springframework.messaging.handler.annotation.Payload
import cafeteria.config.kafka.KafkaProcessor
import cafeteria.external.KakaoMessage
import scala.collection.JavaConverters._
import org.springframework.data.mongodb.core.MongoOperations
import org.springframework.data.mongodb.core.query.Update
import org.springframework.data.mongodb.core.query.Criteria
import org.springframework.data.mongodb.core.query.Query
import org.springframework.data.mongodb.core.FindAndModifyOptions

@Service
class MypageViewHandler {
  
  @Autowired
	private val kakaoService :KakaoService = null
	
  @Autowired
  private val mypageRepository :MypageRepository = null
  
  @Autowired
  private val mongoOperations :MongoOperations = null
  
  def generateSequence (seqName :String) :Long = {
    val query :Query = new Query(Criteria.where("_id").is(seqName));
    val options :FindAndModifyOptions = new FindAndModifyOptions().returnNew(true).upsert(true)
    val update :Update = new Update().inc("seq",1)
    
    val sequence :Option[DatabaseSequence] = Option(mongoOperations.findAndModify(query, update, options, classOf[DatabaseSequence]))
    sequence.getOrElse(InitialSequence(1L)).seq
}
  
  @StreamListener(KafkaProcessor.INPUT)
  def whenOrdered_then_CREATE_1(@Payload ordered :Ordered) {
    try {
      if (ordered.isMe()) {
        
        val mypage :Mypage = new Mypage()
        mypage.id = generateSequence(Mypage.SEQUENCE_NAME)
        mypage.orderId = ordered.id
        mypage.productName = ordered.productName
        mypage.qty = ordered.qty
        mypage.amt = ordered.amt
        mypage.phoneNumber = ordered.phoneNumber
        mypage.status = ordered.status
        
        mypageRepository.save(mypage)
        
        val message :KakaoMessage = new KakaoMessage()
        message.phoneNumber = ordered.phoneNumber
        message.message = s"""Your Order is ${ordered.status}"""
        kakaoService.sendKakao(message)
      }
    } catch {
      case e :Exception => e.printStackTrace()
    }
  }

  @StreamListener(KafkaProcessor.INPUT)
  def whenOrderCanceled_then_UPDATE_1(@Payload orderCanceled :OrderCanceled) {
    try {
      if (orderCanceled.isMe()) {
        mypageRepository.findByOrderId(orderCanceled.id).asScala.foreach(page => {
          page.status = orderCanceled.status
          mypageRepository.save(page)
        })
      
        val message :KakaoMessage = new KakaoMessage()
        message.phoneNumber = orderCanceled.phoneNumber
        message.message = s"""Your Order is ${orderCanceled.status}"""
        kakaoService.sendKakao(message)
      }
    } catch {
      case e :Exception => e.printStackTrace()
    }
  }
  
  @StreamListener(KafkaProcessor.INPUT)
  def whenReceipted_then_UPDATE_2(@Payload receipted :Receipted) {
    try {
      if (receipted.isMe()) {
        mypageRepository.findByOrderId(receipted.orderId).asScala.foreach(page => {
          page.status = receipted.status
          mypageRepository.save(page)
        })
      
        val message :KakaoMessage = new KakaoMessage()
        message.phoneNumber = receipted.phoneNumber
        message.message = s"""Your Order is ${receipted.status}"""
        kakaoService.sendKakao(message)
      }
    } catch {
      case e :Exception => e.printStackTrace()
    }
  }
  
  @StreamListener(KafkaProcessor.INPUT)
  def whenReceipted_then_UPDATE_3(@Payload made :Made) {
    try {
      if (made.isMe()) {
        mypageRepository.findByOrderId(made.orderId).asScala.foreach(page => {
          page.status = made.status
          mypageRepository.save(page)
        })
      
        val message :KakaoMessage = new KakaoMessage()
        message.phoneNumber = made.phoneNumber
        message.message = s"""Your Order is ${made.status}\nCome and Take it, Please!"""
        kakaoService.sendKakao(message)
      }
    } catch {
      case e :Exception => e.printStackTrace()
    }
  }
   
}